### PR TITLE
block:passthrough with multipath device

### PR DESCRIPTION
**Pull request by maxujun**
…path device
ID:1945860

Signed-off-by: Xujun Ma <xuma@redhat.com>

### CODE REVIEW

**Comment by maxujun** [comment]
@qingwangrh Could you help review?

**Comment by qunfang** [comment]
@maxujun " ID：1945860" format is incorrect, probably you were using Chinese input for the ":" symbol.  Please update.

**Comment by maxujun** [comment]
> 
> 
> @maxujun " ID：1945860" format is incorrect, probably you were using Chinese input for the ":" symbol. Please update.

Updated

File: qemu/tests/cfg/passthrough_with_multipath_device.cfg [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    login_timeout = 240
+    start_vm = yes
+    only Linux
+    variants:
+        - multipath_device:
+            pre_command = "yum install device-mapper-multipath -y;mpathconf --enable;systemctl restart multipathd;"
+            pre_command += "modprobe scsi_debug dev_size_mb=5000 num_tgts=1 vpd_use_hostno=0 add_host=2 delay=1 max_luns=2 no_lun_0=1;sleep 5"
+            post_cmd = "multipath -F;sleep 3;modprobe -r scsi_debug"
+            images += " stg0"
+            image_name_stg0 = "/dev/mapper/mpatha"
```
Comment (qingwangrh) [inline qemu/tests/cfg/passthrough_with_multipath_device.cfg:14]: We need to find the mpath instead of fixedsdssds4seeedseseewwww name, due to some hosts have exist other multipath already.

File: qemu/tests/passthrough_with_multipath_device.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    """
+    Test failover path on SCSI passthrough with underlying DM-multipath device.
+
+    Step:
+     1. Build multipath device on host.
+     2. Boot a guest with passthrough path.
+     3. Access guest then do io on the data disk.
+     4. Check vm status.
+     5. Alternately close a path every 10 seconds on host
+     6. Ceck vm status
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:22]: Ceck -> Check

File: qemu/tests/passthrough_with_multipath_device.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    get_mpath_cmd = params.get("get_mpath_cmd")
+    get_mdev_cmd = params.get("get_mdev_cmd")
+    set_path_cmd = params.get("set_path_cmd")
+    post_cmd = params.get("post_cmd")
+    repeat_time = params.get_numeric("repeat_time")
+    session = vm.wait_for_login(timeout=timeout)
+    cmd_dd = "dd if=/dev/zero of=/dev/sdb bs=1M count=300"
+    error_context.context('Do dd writing test on the data disk.',
+                          logging.info)
+    status = session.cmd_status(cmd_dd, timeout=timeout)
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:44]: This command should keep running in guest during switching path on host

**Comment by qingwangrh** [comment]
Please post your test result @maxujun

File: qemu/tests/cfg/passthrough_with_multipath_device.cfg [hunk 4]
```diff
@@ -0,0 +1,24 @@
+- passthrough_with:
```
Comment (qingwangrh) [inline qemu/tests/cfg/passthrough_with_multipath_device.cfg:1]: keep case name same as code file name is better

File: qemu/tests/cfg/passthrough_with_multipath_device.cfg [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    only Linux
+    variants:
+        - multipath_device:
+            pre_command = "yum install device-mapper-multipath -y;mpathconf --enable;systemctl restart multipathd;"
+            pre_command += "modprobe scsi_debug dev_size_mb=5000 num_tgts=1 vpd_use_hostno=0 add_host=2 delay=1 max_luns=2 no_lun_0=1;sleep 5"
+            post_cmd = "multipath -F;sleep 3;modprobe -r scsi_debug"
+            images += " stg0"
+            image_name_stg0 = "/dev/mapper/mpatha"
+            image_format_stg0 = raw
+            image_raw_device_stg0 = yes
```
Comment (qingwangrh) [inline qemu/tests/cfg/passthrough_with_multipath_device.cfg:16]: miss drive_format_stg0=scsi-block ?

**Comment by maxujun** [comment]
(1/1) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.passthrough_with_multipath_device: PASS (609.95 s)
Updated.

**Comment by maxujun** [comment]
@qingwangrh Could you help review again.

**Comment by qingwangrh** [comment]
> (1/1) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.passthrough_with_multipath_device: PASS (609.95 s)
> Updated.

I wonder the test result , it should be failed due to this bug not fixed yet.

File: qemu/tests/passthrough_with_multipath_device.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    repeat_time = params.get_numeric("repeat_time")
+    id = process.getoutput(get_id_cmd).strip()
+    get_mpath_cmd = get_mpath_cmd % (id, id)
+    mpath = process.getoutput(get_mpath_cmd).strip()
+    params["image_name_stg0"] = "/dev/mapper/%s" % mpath
+    params['start_vm'] = 'yes'
+    env_process.process(test, params, env, env_process.preprocess_image,
+                        env_process.preprocess_vm)
+    session = vm.wait_for_login(timeout=timeout)
+    cmd_dd = "dd if=/dev/zero of=/dev/sdb bs=1M count=30000 &"
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:49]: Do you test by manually it have enough time run the io, you may try 
dd if=/dev/zero of=/dev/sdb bs=4k count=1000000 iflag=direct

File: qemu/tests/passthrough_with_multipath_device.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    error_context.context('Do dd writing test on the data disk.',
+                          logging.info)
+    status = session.cmd_status(cmd_dd, timeout=timeout)
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:52]: you may using send line, we do not need to wait the command finished

File: qemu/tests/passthrough_with_multipath_device.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            process.run("multipath -l")
+            time.sleep(1)
+        time.sleep(1)
+    for dev in mdev:
+        process.run(set_path_cmd % ("running", dev))
+
+    if not vm.monitor.verify_status("running"):
+        test.fail("Guest did not run after dd")
+    for dev in mdev:
+        process.run(set_path_cmd % ("offline", dev))
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:81]: does it need to check the vm status?

File: qemu/tests/passthrough_with_multipath_device.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    dev = random.choice(mdev)
+    process.run(set_path_cmd % ("running", dev))
+    session.close()
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:84]: does it need to check the vm status?

**Comment by qingwangrh** [comment]
@YongxueHong Please help to review.thanks

File: qemu/tests/cfg/passthrough_with_multipath_device.cfg [hunk 10]
```diff
@@ -0,0 +1,22 @@
+- passthrough_with_multipath_device:
+    virt_test_type = qemu
+    type = passthrough_with_multipath_device
+    kill_vm_on_error = yes
```
Comment (YongxueHong) [inline qemu/tests/cfg/passthrough_with_multipath_device.cfg:4]: Please replace it with `kill_vm = yes` since `kill_vm_on_error` is dropped from framework.

File: qemu/tests/passthrough_with_multipath_device.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    mpath = process.getoutput(get_mpath_cmd).strip()
+    params["image_name_stg0"] = "/dev/mapper/%s" % mpath
+    params['start_vm'] = 'yes'
+    env_process.process(test, params, env, env_process.preprocess_image,
+                        env_process.preprocess_vm)
+    session = vm.wait_for_login(timeout=timeout)
+    cmd_dd = "dd if=/dev/zero of=/dev/sdb bs=1M count=30000"
+    error_context.context('Do dd writing test on the data disk.',
+                          logging.info)
+    session.sendline(cmd_dd)
```
Comment (YongxueHong) [inline qemu/tests/passthrough_with_multipath_device.py:52]: Suggest that we should add a step to confirm whether the dd process is running if I am afraid there will raise an expected error. For example, there is no space to write data.
BTW, should we add the check dd process step before` if not vm.monitor.verify_status("running"):` to ensure the dd process is running @qingwangrh .

Comment (qingwangrh) [reply qemu/tests/passthrough_with_multipath_device.py:52]: The test point is make sure dd keep running, that is why it need to test cmd_dd = "dd if=/dev/zero of=/dev/sdb bs=1M count=30000" have enough time. but this command look like have issue it will hit have no space to do io. 
I suggest to simply the test step  using proper dd cmd and adding last step to check dd still running .

File: qemu/tests/passthrough_with_multipath_device.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    repeat_time = params.get_numeric("repeat_time")
+    id = process.getoutput(get_id_cmd).strip()
+    get_mpath_cmd = get_mpath_cmd % (id, id)
+    mpath = process.getoutput(get_mpath_cmd).strip()
+    params["image_name_stg0"] = "/dev/mapper/%s" % mpath
+    params['start_vm'] = 'yes'
+    env_process.process(test, params, env, env_process.preprocess_image,
+                        env_process.preprocess_vm)
+    session = vm.wait_for_login(timeout=timeout)
+    cmd_dd = "dd if=/dev/zero of=/dev/sdb bs=1M count=30000"
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:49]: do you test the command manually how long it running ?

File: qemu/tests/passthrough_with_multipath_device.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+            time.sleep(1)
+        time.sleep(1)
+    for dev in mdev:
+        process.run(set_path_cmd % ("running", dev))
+
+    if not vm.monitor.verify_status("running"):
+        test.fail("Guest did not run after dd")
+    for dev in mdev:
+        process.run(set_path_cmd % ("offline", dev))
+    if not vm.monitor.verify_status("running"):
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:82]: wrong condition ? the vm should step in pause?

File: qemu/tests/passthrough_with_multipath_device.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+            process.run(set_path_cmd % ("running", dev))
+            time.sleep(5)
+            process.run("multipath -l")
+            time.sleep(1)
+        time.sleep(1)
+    for dev in mdev:
+        process.run(set_path_cmd % ("running", dev))
+
+    if not vm.monitor.verify_status("running"):
+        test.fail("Guest did not run after dd")
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:79]: do we need to wait_until?

File: qemu/tests/passthrough_with_multipath_device.py (hunk 13 above)
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:82]: do we need to wait_until?

Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:82]: do we need to wait_until?

**Comment by XueqiangWei** [comment]
@maxujun CI check error, indentation is not a multiple of four (comment). Please check it.

**Comment by YongxueHong** [comment]
Please shorten your commit header, the maximin number of characters is 69.
Thanks.

**Comment by maxujun** [comment]
Updated

File: qemu/tests/passthrough_with_multipath_device.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+                               timeout=20):
+        test.fail("Guest did not run after online")
+    for dev in mdev:
+        process.getoutput(set_path_cmd % ("offline", dev))
+    if not utils_misc.wait_for(lambda: vm.monitor.verify_status("paused"),
+                               timeout=20):
+        test.fail("Guest did not pause after offline")
+    dev = random.choice(mdev)
+    process.getoutput(set_path_cmd % ("running", dev))
+    if not utils_misc.wait_for(lambda: vm.monitor.verify_status("running"),
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:85]: Do we need to add resume ?

File: qemu/tests/cfg/passthrough_with_multipath_device.cfg [hunk 16]
```diff
@@ -0,0 +1,23 @@
+- passthrough_with_multipath_device:
+    virt_test_type = qemu
+    type = passthrough_with_multipath_device
+    kill_vm = yes
+    login_timeout = 240
+    start_vm = no
+    only Linux
```
Comment (qingwangrh) [inline qemu/tests/cfg/passthrough_with_multipath_device.cfg:7]: only virtio_scsi ? due to you have hardcode sdb

**Comment by qingwangrh** [comment]
Please post latest test result.

**Comment by maxujun** [comment]
(1/2) Host_RHEL.m8.u5.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.passthrough_with_multipath_device: FAIL: Guest did not run after online (574.09 s)
 (2/2) Host_RHEL.m8.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.passthrough_with_multipath_device: FAIL: Guest did not run after online (575.04 s)

File: qemu/tests/passthrough_with_multipath_device.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+                               timeout=20):
+        vm.monitor.send_args_cmd("c")
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:90]: should before check status ?

File: qemu/tests/cfg/passthrough_with_multipath_device.cfg [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+    images += " stg0"
+    image_format_stg0 = raw
+    drive_format_stg0 = scsi-block
+    image_raw_device_stg0 = yes
+    offline_time = 10
+    repeat_time = 10
+    get_id_cmd = "multipath -l | grep scsi_debug | tr -d "()" | awk 'NR==1{print $2}'"
+    get_mpath_cmd = "multipath -l %s | grep %s | awk 'NR==1{print $1}'"
+    get_mdev_cmd = "lsscsi -ik | grep %s | awk '{print $6}'"
+    get_tdev_cmd = "lsscsi|grep scsi_debug | awk '{print $6}'"
```
Comment (qingwangrh) [inline qemu/tests/cfg/passthrough_with_multipath_device.cfg:21]: There is underlying issue is the device not created by scsi_debug, so far it is acceptable.

File: qemu/tests/passthrough_with_multipath_device.py [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+            process.getoutput(set_path_cmd % ("offline", dev))
+            time.sleep(5)
+            process.getoutput("multipath -l")
+            time.sleep(10)
+            process.getoutput(set_path_cmd % ("running", dev))
+            time.sleep(5)
+            process.getoutput("multipath -l")
+            time.sleep(1)
+        time.sleep(1)
+    for dev in mdev:
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:76]: why put all path running?

File: qemu/tests/passthrough_with_multipath_device.py [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+        process.getoutput(set_path_cmd % ("running", dev))
+    if not utils_misc.wait_for(lambda: vm.monitor.verify_status("running"),
+                               timeout=20):
+        test.fail("Guest did not run after online")
```
Comment (qingwangrh) [inline qemu/tests/passthrough_with_multipath_device.py:80]: suggest to Guest did not run after change path

File: qemu/tests/passthrough_with_multipath_device.py (hunk 19 above)
Comment (maxujun) [reply qemu/tests/passthrough_with_multipath_device.py:76]: It's steps in offline.sh script.

echo "Back to running status for all paths"
for dev in $mdevs; do
  echo running >/sys/block/${dev}/device/state
done

**Comment by maxujun** [comment]
Updated

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=passthrough_with_multipath_device --guestname=RHEL.8.4.0 --clone=no --driveformat=virtio_scsi

(1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.passthrough_with_multipath_device: ERROR: VM creation command failed:    "MALLOC_PERTURB_=1  /usr/libexec/qemu-kvm -S  -name 'avocado-vt-vm1'  -sandbox on  -machine pc  -nodefaults -device VGA,bus=pci.0,addr=0x2 -m 21504  -smp 10,maxcpus=10,cores=5,threads=1,dies=1,sockets=2  -cpu 'Cascadelake-Se... (21.52 s)

Please check log: http://10.73.196.27:8001/job-2021-05-12T05.51-ed54d0b/job.log

**Comment by maxujun** [comment]
Updated

**Comment by maxujun** [comment]
Updated

**Comment by maxujun** [comment]
Updated

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=passthrough_with_multipath_device --guestname=RHEL.8.4.0 --clone=no --driveformat=virtio_scsi
 (1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.passthrough_with_multipath_device: FAIL: Guest did not run after change path (541.09 s)

Test fail is expected 

LGTM

**Review by YongxueHong — APPROVED** [review]
